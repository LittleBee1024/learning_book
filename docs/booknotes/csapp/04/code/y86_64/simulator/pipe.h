#pragma once

#include "./pipe_regs.h"
#include "isa.h"

/* Defined in sim_pipe.cpp */
namespace PIPE
{
   // fetch stage functions
   word_t gen_f_pc();
   word_t gen_f_predPC();
   word_t gen_f_icode();
   word_t gen_f_ifun();
   word_t gen_f_stat();
   word_t gen_need_regids();
   word_t gen_need_valC();
   word_t gen_instr_valid();

   // decocode and writeback stage functions
   word_t gen_d_srcA();
   word_t gen_d_srcB();
   word_t gen_d_dstE();
   word_t gen_d_dstM();
   word_t gen_d_valA();
   word_t gen_d_valB();
   word_t gen_w_dstE();
   word_t gen_w_valE();
   word_t gen_w_dstM();
   word_t gen_w_valM();
   word_t gen_Stat();

   // execute stage functions
   word_t gen_alufun();
   word_t gen_set_cc();
   word_t gen_Bch();
   word_t gen_aluA();
   word_t gen_aluB();
   word_t gen_e_valA();
   word_t gen_e_dstE();

   // memory stage functions
   word_t gen_mem_addr();
   word_t gen_mem_read();
   word_t gen_mem_write();
   word_t gen_m_stat();

   // pipeline control functions for different stages
   word_t gen_F_stall();
   word_t gen_F_bubble();
   word_t gen_D_stall();
   word_t gen_D_bubble();
   word_t gen_E_stall();
   word_t gen_E_bubble();
   word_t gen_M_stall();
   word_t gen_M_bubble();
   word_t gen_W_stall();
   word_t gen_W_bubble();

   /* Intermdiate stage values that must be used by control functions */
   extern byte_t imem_icode;
   extern byte_t imem_ifun;
   extern bool imem_error;
   extern bool instr_valid;
   extern word_t d_regvala;
   extern word_t d_regvalb;
   extern bool dmem_error;

   // export pipeline registers as a global varible
   extern SIM::PipeRegs pipe_regs;
}
