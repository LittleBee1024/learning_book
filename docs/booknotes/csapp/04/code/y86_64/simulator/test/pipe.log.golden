[INFO] Create Pipeline Simulator
[INFO] Loaded 141 bytes code

Cycle 0. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x0
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 0, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 1. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0xa
D: instr = irmovq, rA = ----, rB = %rsp, valC = 0x200, valP = 0xa, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 2. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x38
D: instr = call, rA = ----, rB = ----, valC = 0x38, valP = 0x13, Stat = OK
E: instr = irmovq, valC = 0x200, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = %rsp, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 3. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x42
D: instr = irmovq, rA = ----, rB = %rdi, valC = 0x18, valP = 0x42, Stat = OK
E: instr = call, valC = 0x38, valA = 0x13, valB = 0x200
   srcA = ----, srcB = %rsp, dstE = %rsp, dstM = ----, Stat = OK
M: instr = irmovq, Cnd = 1, valE = 0x200, valA = 0x0
   dstE = %rsp, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 4. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x4c
D: instr = irmovq, rA = ----, rB = %rsi, valC = 0x4, valP = 0x4c, Stat = OK
E: instr = irmovq, valC = 0x18, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = %rdi, dstM = ----, Stat = OK
M: instr = call, Cnd = 1, valE = 0x1f8, valA = 0x13
   dstE = %rsp, dstM = ----, Stat = OK
W: instr = irmovq, valE = 0x200, valM = 0x0, dstE = %rsp, dstM = ----, Stat = OK

Cycle 5. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = call, rA = ----, rB = ----, valC = 0x56, valP = 0x55, Stat = OK
E: instr = irmovq, valC = 0x4, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = %rsi, dstM = ----, Stat = OK
M: instr = irmovq, Cnd = 1, valE = 0x18, valA = 0x0
   dstE = %rdi, dstM = ----, Stat = OK
W: instr = call, valE = 0x1f8, valM = 0x0, dstE = %rsp, dstM = ----, Stat = OK

Cycle 6. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x60
D: instr = irmovq, rA = ----, rB = %r8, valC = 0x8, valP = 0x60, Stat = OK
E: instr = call, valC = 0x56, valA = 0x55, valB = 0x1f8
   srcA = ----, srcB = %rsp, dstE = %rsp, dstM = ----, Stat = OK
M: instr = irmovq, Cnd = 1, valE = 0x4, valA = 0x0
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = irmovq, valE = 0x18, valM = 0x0, dstE = %rdi, dstM = ----, Stat = OK

Cycle 7. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6a
D: instr = irmovq, rA = ----, rB = %r9, valC = 0x1, valP = 0x6a, Stat = OK
E: instr = irmovq, valC = 0x8, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = %r8, dstM = ----, Stat = OK
M: instr = call, Cnd = 1, valE = 0x1f0, valA = 0x55
   dstE = %rsp, dstM = ----, Stat = OK
W: instr = irmovq, valE = 0x4, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 8. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6c
D: instr = xorq, rA = %rax, rB = %rax, valC = 0x0, valP = 0x6c, Stat = OK
E: instr = irmovq, valC = 0x1, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = %r9, dstM = ----, Stat = OK
M: instr = irmovq, Cnd = 1, valE = 0x8, valA = 0x0
   dstE = %r8, dstM = ----, Stat = OK
W: instr = call, valE = 0x1f0, valM = 0x0, dstE = %rsp, dstM = ----, Stat = OK

Cycle 9. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x6e
D: instr = andq, rA = %rsi, rB = %rsi, valC = 0x0, valP = 0x6e, Stat = OK
E: instr = xorq, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = %rax, srcB = %rax, dstE = %rax, dstM = ----, Stat = OK
M: instr = irmovq, Cnd = 1, valE = 0x1, valA = 0x0
   dstE = %r9, dstM = ----, Stat = OK
W: instr = irmovq, valE = 0x8, valM = 0x0, dstE = %r8, dstM = ----, Stat = OK

Cycle 10. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = jmp, rA = ----, rB = ----, valC = 0x87, valP = 0x77, Stat = OK
E: instr = andq, valC = 0x0, valA = 0x4, valB = 0x4
   srcA = %rsi, srcB = %rsi, dstE = %rsi, dstM = ----, Stat = OK
M: instr = xorq, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = %rax, dstM = ----, Stat = OK
W: instr = irmovq, valE = 0x1, valM = 0x0, dstE = %r9, dstM = ----, Stat = OK

Cycle 11. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90, Stat = OK
E: instr = jmp, valC = 0x87, valA = 0x77, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = andq, Cnd = 0, valE = 0x4, valA = 0x4
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = xorq, valE = 0x0, valM = 0x0, dstE = %rax, dstM = ----, Stat = OK

Cycle 12. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81, Stat = OK
E: instr = jne, valC = 0x77, valA = 0x90, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = jmp, Cnd = 1, valE = 0x0, valA = 0x77
   dstE = ----, dstM = ----, Stat = OK
W: instr = andq, valE = 0x4, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 13. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x18
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10, Stat = OK
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x90
   dstE = ----, dstM = ----, Stat = OK
W: instr = jmp, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 14. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = mrmovq, Cnd = 1, valE = 0x18, valA = 0x0
   dstE = ----, dstM = %r10, Stat = OK
W: instr = jne, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 15. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85, Stat = OK
E: instr = addq, valC = 0x0, valA = 0xd000d000d, valB = 0x0
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = mrmovq, valE = 0x18, valM = 0xd000d000d, dstE = ----, dstM = %r10, Stat = OK

Cycle 16. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87, Stat = OK
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x18
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0xd000d000d, valA = 0xd000d000d
   dstE = %rax, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 17. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90, Stat = OK
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x4
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0x20, valA = 0x8
   dstE = %rdi, dstM = ----, Stat = OK
W: instr = addq, valE = 0xd000d000d, valM = 0x0, dstE = %rax, dstM = ----, Stat = OK

Cycle 18. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81, Stat = OK
E: instr = jne, valC = 0x77, valA = 0x90, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = subq, Cnd = 0, valE = 0x3, valA = 0x1
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = addq, valE = 0x20, valM = 0x0, dstE = %rdi, dstM = ----, Stat = OK

Cycle 19. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x20
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10, Stat = OK
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x90
   dstE = ----, dstM = ----, Stat = OK
W: instr = subq, valE = 0x3, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 20. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = mrmovq, Cnd = 1, valE = 0x20, valA = 0x0
   dstE = ----, dstM = %r10, Stat = OK
W: instr = jne, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 21. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85, Stat = OK
E: instr = addq, valC = 0x0, valA = 0xc000c000c0, valB = 0xd000d000d
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = mrmovq, valE = 0x20, valM = 0xc000c000c0, dstE = ----, dstM = %r10, Stat = OK

Cycle 22. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87, Stat = OK
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x20
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0xcd00cd00cd, valA = 0xc000c000c0
   dstE = %rax, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 23. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90, Stat = OK
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x3
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0x28, valA = 0x8
   dstE = %rdi, dstM = ----, Stat = OK
W: instr = addq, valE = 0xcd00cd00cd, valM = 0x0, dstE = %rax, dstM = ----, Stat = OK

Cycle 24. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81, Stat = OK
E: instr = jne, valC = 0x77, valA = 0x90, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = subq, Cnd = 0, valE = 0x2, valA = 0x1
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = addq, valE = 0x28, valM = 0x0, dstE = %rdi, dstM = ----, Stat = OK

Cycle 25. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x28
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10, Stat = OK
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x90
   dstE = ----, dstM = ----, Stat = OK
W: instr = subq, valE = 0x2, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 26. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = mrmovq, Cnd = 1, valE = 0x28, valA = 0x0
   dstE = ----, dstM = %r10, Stat = OK
W: instr = jne, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 27. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85, Stat = OK
E: instr = addq, valC = 0x0, valA = 0xb000b000b00, valB = 0xcd00cd00cd
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = mrmovq, valE = 0x28, valM = 0xb000b000b00, dstE = ----, dstM = %r10, Stat = OK

Cycle 28. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87, Stat = OK
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x28
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0xbcd0bcd0bcd, valA = 0xb000b000b00
   dstE = %rax, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 29. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90, Stat = OK
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x2
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0x30, valA = 0x8
   dstE = %rdi, dstM = ----, Stat = OK
W: instr = addq, valE = 0xbcd0bcd0bcd, valM = 0x0, dstE = %rax, dstM = ----, Stat = OK

Cycle 30. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81, Stat = OK
E: instr = jne, valC = 0x77, valA = 0x90, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = subq, Cnd = 0, valE = 0x1, valA = 0x1
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = addq, valE = 0x30, valM = 0x0, dstE = %rdi, dstM = ----, Stat = OK

Cycle 31. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = mrmovq, valC = 0x0, valA = 0x0, valB = 0x30
   srcA = ----, srcB = %rdi, dstE = ----, dstM = %r10, Stat = OK
M: instr = jne, Cnd = 1, valE = 0x0, valA = 0x90
   dstE = ----, dstM = ----, Stat = OK
W: instr = subq, valE = 0x1, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 32. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = addq, rA = %r10, rB = %rax, valC = 0x0, valP = 0x83, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = mrmovq, Cnd = 1, valE = 0x30, valA = 0x0
   dstE = ----, dstM = %r10, Stat = OK
W: instr = jne, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 33. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x85
D: instr = addq, rA = %r8, rB = %rdi, valC = 0x0, valP = 0x85, Stat = OK
E: instr = addq, valC = 0x0, valA = 0xa000a000a000, valB = 0xbcd0bcd0bcd
   srcA = %r10, srcB = %rax, dstE = %rax, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = mrmovq, valE = 0x30, valM = 0xa000a000a000, dstE = ----, dstM = %r10, Stat = OK

Cycle 34. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x87
D: instr = subq, rA = %r9, rB = %rsi, valC = 0x0, valP = 0x87, Stat = OK
E: instr = addq, valC = 0x0, valA = 0x8, valB = 0x30
   srcA = %r8, srcB = %rdi, dstE = %rdi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0xabcdabcdabcd, valA = 0xa000a000a000
   dstE = %rax, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 35. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x77
D: instr = jne, rA = ----, rB = ----, valC = 0x77, valP = 0x90, Stat = OK
E: instr = subq, valC = 0x0, valA = 0x1, valB = 0x1
   srcA = %r9, srcB = %rsi, dstE = %rsi, dstM = ----, Stat = OK
M: instr = addq, Cnd = 1, valE = 0x38, valA = 0x8
   dstE = %rdi, dstM = ----, Stat = OK
W: instr = addq, valE = 0xabcdabcdabcd, valM = 0x0, dstE = %rax, dstM = ----, Stat = OK

Cycle 36. CC=Z=0 S=0 O=0, Stat=OK
F: predPC = 0x81
D: instr = mrmovq, rA = %r10, rB = %rdi, valC = 0x0, valP = 0x81, Stat = OK
E: instr = jne, valC = 0x77, valA = 0x90, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = OK
M: instr = subq, Cnd = 0, valE = 0x0, valA = 0x1
   dstE = %rsi, dstM = ----, Stat = OK
W: instr = addq, valE = 0x38, valM = 0x0, dstE = %rdi, dstM = ----, Stat = OK

Cycle 37. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x83
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = jne, Cnd = 0, valE = 0x0, valA = 0x90
   dstE = ----, dstM = ----, Stat = OK
W: instr = subq, valE = 0x0, valM = 0x0, dstE = %rsi, dstM = ----, Stat = OK

Cycle 38. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x91
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x91, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = jne, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = OK

Cycle 39. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x91
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = ret, valC = 0x0, valA = 0x1f0, valB = 0x1f0
   srcA = %rsp, srcB = %rsp, dstE = %rsp, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 40. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x91
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = ret, Cnd = 1, valE = 0x1f8, valA = 0x1f0
   dstE = %rsp, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 41. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x91
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = ret, valE = 0x1f8, valM = 0x55, dstE = %rsp, dstM = ----, Stat = OK

Cycle 42. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = ret, rA = ----, rB = ----, valC = 0x0, valP = 0x56, Stat = OK
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 43. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = ret, valC = 0x0, valA = 0x1f8, valB = 0x1f8
   srcA = %rsp, srcB = %rsp, dstE = %rsp, dstM = ----, Stat = OK
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 44. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = ret, Cnd = 1, valE = 0x200, valA = 0x1f8
   dstE = %rsp, dstM = ----, Stat = OK
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 45. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x56
D: instr = nop, rA = ----, rB = ----, valC = 0x0, valP = 0x0, Stat = BUBBLE
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = ret, valE = 0x200, valM = 0x13, dstE = %rsp, dstM = ----, Stat = OK

Cycle 46. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x14
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x14, Stat = HALT
E: instr = nop, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = BUBBLE
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 47. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x15
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x15, Stat = HALT
E: instr = halt, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = HALT
M: instr = nop, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 48. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x16
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x16, Stat = HALT
E: instr = halt, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = HALT
M: instr = halt, Cnd = 1, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = HALT
W: instr = nop, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = BUBBLE

Cycle 49. CC=Z=1 S=0 O=0, Stat=OK
F: predPC = 0x17
D: instr = halt, rA = ----, rB = ----, valC = 0x0, valP = 0x17, Stat = HALT
E: instr = halt, valC = 0x0, valA = 0x0, valB = 0x0
   srcA = ----, srcB = ----, dstE = ----, dstM = ----, Stat = HALT
M: instr = nop, Cnd = 0, valE = 0x0, valA = 0x0
   dstE = ----, dstM = ----, Stat = BUBBLE
W: instr = halt, valE = 0x0, valM = 0x0, dstE = ----, dstM = ----, Stat = HALT

After 50 cycles, the state becomes HALT

Changes to registers:
%rax:	0x0000000000000000	0x0000abcdabcdabcd
%rsp:	0x0000000000000000	0x0000000000000200
%rdi:	0x0000000000000000	0x0000000000000038
%r8:	0x0000000000000000	0x0000000000000008
%r9:	0x0000000000000000	0x0000000000000001
%r10:	0x0000000000000000	0x0000a000a000a000
Changes to memory:
0x01f0:	0x0000000000000000	0x0000000000000055
0x01f8:	0x0000000000000000	0x0000000000000013
